libs/partners/google-genai/langchain_google_genai/chat_models.py

```diff
@@ -332,14 +333,90 @@ def _parse_chat_history(
     return messages
 
 
-def _parts_to_content(parts: List[genai.types.PartType]) -> Union[List[dict], str]:
+def _retrieve_function_call_response(
```

Comment:
a lot of stuff is going here, it might be a good idea to move this into a separate module (similar to https://github.com/langchain-ai/langchain/blob/master/libs/partners/google-vertexai/langchain_google_vertexai/functions_utils.py) and implement a function that converts a `langchain_core.tools.BaseTool` into `glm.Tool`

Reply:
for future followup